mlx/array.h

```diff
@@ -93,6 +95,21 @@ class array {
     return array_desc_->shape.size();
   };
 
+  template <typename T>
+  static int check_shape_dim(const T dim) {
+    constexpr bool is_signed = std::numeric_limits<T>::is_signed;
+    using U = std::conditional_t<is_signed, ssize_t, size_t>;
+    constexpr U min = static_cast<U>(std::numeric_limits<int>::min());
+    constexpr U max = static_cast<U>(std::numeric_limits<int>::max());
+
+    if ((is_signed && dim < min) || dim > max) {
+      throw std::invalid_argument(
+          "Shape dimension falls outside supported `int` range.");
+    }
+
+    return static_cast<int>(dim);
+  }
```

Comment:
Also I don't think we need this to be tepmlated (?). I would just have it take a `size_t` and simplify things a bit.

Reply:
Templated because `shape` returns `ssize_t` whereas `len` returns `size_t`. The reason I didn't place in `python/sr/array.cpp` was so that I could cover additional edge cases with C++ test cases.

Reply:
Ok fair enough. I prefer it not be in `array.h`. You could put it in `mlx/utils.h`.

As for the template, I think you mean `shape` return `int`? It will be an always positive int, so you could cast it, but I don't feel too strongly about it.

Reply:
Moved to `mlx/utils.h` and kept template as `py::shape` returns `ssize_t` and `py::len` returns `size_t`.